[SE-3329] Expose banner image url in course API

gabor-boros opened the pull request:
This PR extends the course overview and detail APIs to include (both relative and absolute) banner image url, which allows content publishers to specify a banner image and then reuse that from a custom UI/integration.

**Dependencies**: None

**Screenshots**:

![Screenshot_2020-11-12 Course Detail – Django REST framework](https://user-images.githubusercontent.com/19173947/99000523-b382f600-2539-11eb-84ea-61526843c73a.png)
![Screenshot_2020-11-12 Course List – Django REST framework](https://user-images.githubusercontent.com/19173947/99000525-b4b42300-2539-11eb-8379-6d219470e72b.png)


**Sandbox URL**: https://pr25045.sandbox.opencraft.hosting/

**Merge deadline**: Preferably 29th of September

**Testing instructions**:

1. Open edX Studio from **Content -> Files & Uploads**, then copy the file name
2. Paste the file name into `Course Banner Image` field into **Settings -> Advanced Settings** page
3. Open `<URL>/api/courses/v1/courses/` to check all the courses, where URL is the sandbox LMS URL
4. Open `<URL>/api/courses/v1/courses/<course_id>` to see all the courses, where URL is the sandbox LMS URL and course_id is the course id

**Author notes and concerns**:

1. The `_AbsolutMediaSerializer` is using a private attribute from `AbsoluteURLField`, but I found no other way to avoid code duplication and have the same absolute url conversion functionality, since `AbsoluteURLField` requires the request object
   Currently looking for ways to fix it.

**Reviewers**
- [x] @toxinu 
- [ ] @bradenmacdonald 

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
```

edx-webhook commented:
Thanks for the pull request, @gabor-boros! I've created [OSPR-4994](https://openedx.atlassian.net/browse/OSPR-4994) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.



<!-- comment:external_pr -->
<!-- jenkins ok to test -->

toxinu commented:
@gabor-boros Code reviewed and I go through testing instructions. Please ping me when you want another review 👍🏻

gabor-boros commented:
Then I'm pinging you @toxinu

natabene commented:
@gabor-boros Thank you for your contribution. Please let me know once this is ready for our review.

toxinu commented:
@gabor-boros @natabene Reviewed and approved 👍🏻

gabor-boros commented:
@natabene It is ready from our side 😊

gabor-boros commented:
@natabene do I need to do anything else related to this PR? ☺️

natabene commented:
@gabor-boros No, just wait till we can get to it.

bradenmacdonald commented on lms/djangoapps/course_api/serializers.py:
```diff
@@ -32,6 +32,38 @@ def get_uri(self, course_overview):
+        self.context = serializer_field.context
+        return super(self).__call__(serializer_field)
+
+    url = serializers.SerializerMethodField(source="*")
```
Having two fields named `uri` and `url` is confusing. Also, DRF has [special treatment](https://www.django-rest-framework.org/api-guide/serializers/#serializer_url_field) for fields named `url`. Can we call this `uri_absolute` ?

gabor-boros replied:
This is a bit trickier situation in terms of keeping conventions. 

Although I see your point, since URL represents a reference to a web resource located on the network and URI tells the resource identifier, I'm not sure that we should change these conventions in favor to comply with DRF's implementation. Also, DRF's URL filed is only accessible when the serializer is using `HyperlinkedModelSerializer` as its base. Other than that, `uri_absolute` wouldn't conflict with DRF's `url` field anyway.

As this serializer would possibly never use `HyperlinkedModelSerializer` based on its nature, what would you think about leaving `url` and `uri` here, since it wouldn't conflict with the ModelSerializer and we could keep the general naming conventions. 😊

bradenmacdonald replied:
If you can show me anywhere else in edx-platform that uses `uri` for relative URL and `url` for absolute URL, we can keep that, but otherwise I think it's introducing a new confusing convention. (Well, the real problem here is that `uri` should never have been used for relative paths in the first place; without a scheme part and a `:` it's neither a URL nor URI by any definition). I guess I've changed my mind and now my preference would be to put the absolute URL in `uri` and put the relative path or other identifier in `path` or `id`.

BTW are you 100% sure we need the relative URL to be returned for this use case? By far the cleanest solution is to remove the requirement for returning the relative path.

gabor-boros replied:
As we discussed, I'm renaming the field fro `url` to `uri_absolute`.

bradenmacdonald commented on lms/djangoapps/course_api/serializers.py:
```diff
@@ -32,6 +32,38 @@ def get_uri(self, course_overview):
+        # In order to use the AbsoluteURLField to have the same
+        # behaviour what ImageSerializer provides, we need to set
+        # the request for the field
+        field._context = {"request": self.context.get("request")}
```
I believe this hack is required because you're declaring the field within this method, instead of as a member of this serializer.

I think that if you just declare the field like this:

```python
class _AbsolutMediaSerializer(_MediaSerializer):
    requires_context = True
    url = AbsoluteURLField(source="get_url")

    def get_url(self):
        ...
```
then it should work without `context` hacks.

gabor-boros replied:
Unfortunately, this cannot be done, because the `BaseSerializer` does not set the `self.initial_data` if data argument was empty. Although serializers has a `get_initial` method, the method will return an empty `OrderedDict` since the serializer did not receive data at all. So we need to pass the context around, but I'm opened to other ideas as well. When I did a research in this topic when I opened the PR, I did not find any other solution which is cleaner.

bradenmacdonald commented on openedx/core/djangoapps/content/course_overviews/tests/test_course_overviews.py:
```diff
@@ -789,6 +789,25 @@ def test_cdn_with_external_image(self, modulestore_type):
+            modified_url = overview.apply_cdn_to_url(start_url)
+
+            self.assertNotEqual(start_url, modified_url)
+            self.assertTrue(modified_url.startswith(expected_cdn_url))
```
This test has the same docstring as the previous test and it's hard to tell the difference. It took me a while to figure out that this is testing a slightly different method. Plus it doesn't actually test what it says in the docstring, because this test case doesn't test the case where `start_url` is already an absolute URL.

Maybe combine the tests, or change the docstring to say "Test the `apply_cdn_to_url` method directly when used with a single URL"

gabor-boros replied:
Oh, it seems I overlooked the docstring here. Thanks for pointing out! I'll fix this as well

bradenmacdonald replied:
I think you forgot to address this review comment ^

gabor-boros replied:
Oh, you are totally right. I'm doing it just right now. Sorry about that.

bradenmacdonald reviewed (changes requested):
@gabor-boros Overall seems good, but I have a few requested cleanups.

gabor-boros commented on openedx/core/djangoapps/content/course_overviews/models.py:
```diff
@@ -728,6 +730,21 @@ def closest_released_language(self):
 
+    def apply_cdn_to_url(self, image_url):
+        """
+        Given a resolution -> url, return a copy with CDN applied.
```
👍 Rephrasing the misleading docstring

gabor-boros commented on openedx/core/djangoapps/content/course_overviews/models.py:
```diff
@@ -86,6 +86,7 @@ class Meta(object):
     announcement = DateTimeField(null=True)
 
     # URLs
+    banner_image_url = TextField(null=True)
```
The only reason I used `null=True` to keep the "convention" seen for the other fields, but I'll adjust the code to use a `default=""` instead as the django convention recommends as well.

bradenmacdonald replied:
I see a mix in here where some fields have it and some don't. Since the django convention is well-known, I'm wondering if there's a reason behind it. Unfortunately I don't see anything explained in the comments. Perhaps we should add a comment saying "Not allowing null per django convention; not sure why many TextFields in this model do allow null"

bradenmacdonald replied:
Also it's a good idea to review the migration checklist at the end of https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations

gabor-boros replied:
I checked the page, though I'm lack of access to some resources linked in the page. Thank you for the  reading 😇 
It's a good idea about the comment. I'll add that too.

gabor-boros replied:
Also, I adjusted the migration file (and its name) do you have any other suggestions regarding migrations?

bradenmacdonald replied:
Migration looks good to me, thanks.

gabor-boros commented on openedx/core/djangoapps/content/course_overviews/models.py:
```diff
@@ -196,6 +197,7 @@ def _create_or_update(cls, course):
         course_overview.advertised_start = course.advertised_start
         course_overview.announcement = course.announcement
 
+        course_overview.banner_image_url = CourseDetails.fetch_banner_image_url(course.id)
```
You are totally right, the `course_image_url` could be used as well here, replacing it.

gabor-boros commented:
@bradenmacdonald I just adjusted most of the comments and updated the images in the description. Could you please give it another review round?

bradenmacdonald commented:
Thanks @gabor-boros. Looking good now. Please address the remaining issue with the test/docstring, get the tests passing, and squash down to a single commit with a useful commit message. I will then approve.

Also, can you link me directly to an example on the sandbox where I can see the output? And in future, please use sandboxes from Ocim prod, not Ocim stage, since the stage ones don't have valid HTTPS certificates.

gabor-boros commented:
Thank you for the review @bradenmacdonald!

I did adjust the comments, squashed the commits and replaced the sandbox URL in the PR description besides made sure that tests are passing.

bradenmacdonald reviewed (approved):
:+1:

- [x] I tested this: at https://pr25045.sandbox.opencraft.hosting/api/courses/v1/courses/
- [x] I read through the code
- [x] I checked for accessibility issues: n/a
- [x] Includes documentation: docstrings etc. yes

edx-webhook commented:
@gabor-boros 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/25045&entry.1484655318=https://github.com/gabor-boros&entry.752974735=2020-09-22+15:55&entry.1917517419=2020-11-17+16:08&entry.2133058324=Yes) so we can improve your experience in the future.

nasthagiri commented on openedx/core/djangoapps/content/course_overviews/models.py:
```diff
@@ -86,6 +86,8 @@ class Meta(object):
 
     # URLs
+    # Not allowing null per django convention; not sure why many TextFields in this model do allow null
+    banner_image_url = TextField()
```
Thank you for this platform enhancement.
Where may I find documentation on how a "banner image" differs from a "course image"?

bradenmacdonald replied:
> Where may I find documentation on how a "banner image" differs from a "course image"?

@nasthagiri I see some context here, but not sure that it qualifies as "documentation" : https://github.com/edx/edx-platform/blob/7afee25ce2150443eefe0ae60b81f3e4f6e61900/common/lib/xmodule/xmodule/course_module.py#L589-L609

nasthagiri replied:
From a consumer perspective, it would be useful to know when the difference between the 2 images. Perhaps it would help to describe where each image is displayed?

bradenmacdonald replied:
ping @gabor-boros - can you answer that question ^ ?

gabor-boros replied:
@nasthagiri & @bradenmacdonald  Apologies for replying that late, I probably missed the notification about this comment. As far as I can see, the `banner_image` is [part of](https://github.com/edx/edx-platform/blob/9caf1e51e99d9f2cc6d1d649d4861ef11b3f3df1/cms/templates/settings.html#L449-L515) the `extended_course_details`. The original addition was part of https://github.com/edx/edx-platform/commit/f7f281b65b2617b3b5bf599dcc18e67ddf89acb0 as far as I can see, but I cannot find more information about that either.

From consumer perspective - in my mind - the difference is that the course image is kind of a thumbnail of a course while the banner image is like a cover image. So something like this:

![Screenshot 2020-12-08 at 17 24 17](https://user-images.githubusercontent.com/19173947/101510428-369d3d80-397a-11eb-909f-d1bd50d73832.png)
